### PR TITLE
FOGL-1213 : Password rotation support

### PR DESCRIPTION
**Pull request by mshariq-nerd**
## What do these changes do?
1) Password rotation policy applied for users.
2) Added reset password screen
3) `Update role` and `reset password` option added in user management section  
4) User profile moved into navigation bar dropdown 
5) Api endpoint fixed for admin 

## Are there changes in behavior for the user?
Yes, For users

## Related issue number
[FOGL-1213](https://scaledb.atlassian.net/browse/FOGL-1213)

## Checklist

- [ ] Code is well written w.r.t. best practices and proper comments
- [ ] Unit tests for the changes exist
- [ ] e2e tests for the changes exist
- [ ] Documentation reflects the changes including changelog
- [ ] Build works with nginx-light deployment
- [ ] Build works with Docker, if any change on docker related config files

### CODE REVIEW

File: src/app/login/login.component.ts [hunk 1]
```diff
@@ -50,7 +50,14 @@ export class LoginComponent implements OnInit {
                     this.ngProgress.done();
                     if (error.status === 0) {
                         console.log('service down', error);
-                    } else {
+                    } else if (error.status === 401) {
+                        if (error.statusText.toUpperCase().indexOf('PASSWORD') >= 0 && error.statusText.toUpperCase().indexOf('EXPIRED') >= 0) {
+                            console.log("text", error.statusText);
+                            this.router.navigate(['/reset-password'], { queryParams: { username: this.model.username } });
+                            this.alertService.error(error.statusText);
```
Comment (praveen-garg) [inline src/app/login/login.component.ts]: keep this out of `if (error.statusText.toUpperCase()` block to know what is happening

File: src/app/login/login.component.ts [hunk 2]
```diff
@@ -50,7 +50,14 @@ export class LoginComponent implements OnInit {
                     this.ngProgress.done();
                     if (error.status === 0) {
                         console.log('service down', error);
-                    } else {
+                    } else if (error.status === 401) {
+                        if (error.statusText.toUpperCase().indexOf('PASSWORD') >= 0 && error.statusText.toUpperCase().indexOf('EXPIRED') >= 0) {
+                            console.log("text", error.statusText);
```
Comment (praveen-garg) [inline src/app/login/login.component.ts]: remove log

File: src/app/services/user.service.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { environment } from '../../environments/environment';
 export class UserService {
 
   private USER_URL = environment.BASE_URL + 'user';
+  private ADMIN_BASE_URL = environment.BASE_URL + 'admin';
```
Comment (praveen-garg) [inline src/app/services/user.service.ts]: ADMIN_BASE_URL > ADMIN_URL

File: src/app/user-management/reset-password/reset-password.component.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  ngOnInit() {
+  }
+
+  public resetUserForm(form: NgForm) {
+    form.controls["oldPassword"].reset();
+    form.controls["password"].reset();
+    form.controls["confirmPassword"].reset();
+  }
+
+  resetPassword(form: NgForm) {
```
Comment (praveen-garg) [inline src/app/user-management/reset-password/reset-password.component.ts]: is this public method or protected or private

File: src/app/user-management/update-user/update-user.component.html [hunk 5]
```diff
@@ -2,12 +2,13 @@
   <div class="modal-background"></div>
   <div class="modal-card">
     <header class="modal-card-head">
-      <p class="modal-card-title">Update User</p>
+      <p *ngIf="showRoleSection" class="modal-card-title">Update user role</p>
+      <p *ngIf="!showRoleSection" class="modal-card-title">Update user password</p>
```
Comment (praveen-garg) [inline src/app/user-management/update-user/update-user.component.html]: what is `Update user password` ? toggling?

File: src/app/user-management/user-management.component.html [hunk 6]
```diff
@@ -21,7 +21,7 @@ <h6 class="is-6 has-text-weight-semibold">
           </ul>
           <div>
             <a class="is-light" (click)="openCreateUserModal()">
-              <i class="fa fa-user" aria-hidden="true"></i>&nbsp; Add new member&nbsp;
+              <i class="fa fa-user" aria-hidden="true"></i>&nbsp; Add user&nbsp;
```
Comment (praveen-garg) [inline src/app/user-management/user-management.component.html]: Add User

File: src/app/user-management/user-management.component.html [hunk 7]
```diff
@@ -45,8 +45,13 @@ <h6 class="is-6 has-text-weight-semibold">
...
               <td *ngIf='user.userId != uid'>
-                <a (click)="openUpdateUserModal(user)" class="button is-text">
-                  <i class="fa fa-pencil-square-o" aria-hidden="true"></i>
+                <a (click)="openUpdateUserModal(user, 'role')" class="button is-text">
+                  change role
+                </a>
+              </td>
+              <td *ngIf='user.userId != uid'>
+                <a (click)="openUpdateUserModal(user, 'password')" class="button is-text">
+                  change password
```
Comment (praveen-garg) [inline src/app/user-management/user-management.component.html]: reset password

File: src/app/user-management/user-management.component.ts [hunk 8]
```diff
@@ -72,6 +72,7 @@ export class UserManagementComponent implements OnInit {
             })
           })
           this.userRecord = users;
+          this.userRecord.sort();
```
Comment (praveen-garg) [inline src/app/user-management/user-management.component.ts]: or just assign `users.sort()`

File: src/app/user-management/user-profile/user-profile.component.ts [hunk 9]
```diff
@@ -70,14 +70,35 @@ export class UserProfileComponent implements OnInit {
...
+  public toggleModal(isOpen: Boolean) {
+    let userProfileModal = <HTMLDivElement>document.getElementById('user_profile_modal');
+    if (isOpen) {
+      userProfileModal.classList.add('is-active');
+      return;
+    }
+    userProfileModal.classList.remove('is-active');
+  }
+
+  changePassword(form: NgForm, userName) {
```
Comment (praveen-garg) [inline src/app/user-management/user-profile/user-profile.component.ts]: access modifier missing

File: src/app/user-management/user-profile/user-profile.component.ts [hunk 10]
```diff
@@ -70,14 +70,35 @@ export class UserProfileComponent implements OnInit {
...
+    let passwordPayload: any = {
+      current_password: form.controls["oldPassword"].value,
+      new_password: form.controls["password"].value
```
Comment (praveen-garg) [inline src/app/user-management/user-profile/user-profile.component.ts]: can we make it consistent on FORM as well?

File: src/app/user-management/update-user/update-user.component.html (hunk 5 above)
Comment (mshariq-nerd) [reply src/app/user-management/update-user/update-user.component.html]: toggling of card title based on button action  `reset password` or `change role`

File: src/app/user-management/user-profile/user-profile.component.html [hunk 11]
```diff
@@ -11,66 +11,79 @@
...
+  <div class="modal-card">
+    <header class="modal-card-head">
+      <p class="modal-card-title">Reset Password</p>
+      <button class="delete" aria-label="close" (click)="toggleModal(false)"></button>
+    </header>
+    <section class="modal-card-body">
+      <form name="form" (ngSubmit)="f.form.valid && changePassword(f, userRecord.userName)" #f="ngForm" novalidate>
+        <div class="columns">
+          <div class="column is-6">
+            <input type="password" class="input" placeholder="old password" name='currentPassword' [(ngModel)]="userRecord.currentPassword" #currentPassword="ngModel" required/>
```
Comment (praveen-garg) [inline src/app/user-management/user-profile/user-profile.component.html]: placeholder="old password" should be placeholder="current password"

File: src/app/user-management/reset-password/reset-password.component.html [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  <ng-progress [positionUsing]="'marginLeft'" [minimum]="0.15" [maximum]="1" [speed]="200" [showSpinner]="true" [direction]="'leftToRightIncreased'"
+    [color]="'#1B95E0'" [trickleSpeed]="300" [thick]="true" [ease]="'linear'">
+  </ng-progress>
+  <div class="has-text-centered">
+    <div class="column is-4 is-offset-4">
+      <div class="box">
+        <form name="form" (ngSubmit)="f.form.valid && resetPassword(f)" #f="ngForm" novalidate>
+          <div class="field">
+              <div class="control has-icons-left">
+                  <input type="password" class="input" placeholder="old password" name='oldPassword' [(ngModel)]="userRecord.oldPassword" #oldPassword="ngModel"
```
Comment (praveen-garg) [inline src/app/user-management/reset-password/reset-password.component.html]: placeholder="current password"
oldPassword > currentPassword

**Pull request merged**